contracts/services/astroport-withdrawer/src/astroport_cw20.rs

```diff
@@ -0,0 +1,72 @@
+use cosmwasm_std::{to_json_binary, Addr, Coin, CosmosMsg, DepsMut};
+use cw20::{BalanceResponse, Cw20ExecuteMsg};
+use valence_astroport_utils::astroport_cw20_lp_token::{Asset, PairInfo};
+use valence_service_utils::error::ServiceError;
+
+use crate::msg::Config;
+
+pub fn query_liquidity_token(deps: &DepsMut, cfg: &Config) -> Result<Addr, ServiceError> {
+    let pair_info: PairInfo = deps.querier.query_wasm_smart(
+        cfg.pool_addr.clone(),
+        &valence_astroport_utils::astroport_cw20_lp_token::PoolQueryMsg::Pair {},
+    )?;
+
+    Ok(pair_info.liquidity_token)
+}
+
+pub fn create_withdraw_liquidity_msgs(
+    deps: &DepsMut,
+    cfg: &Config,
+) -> Result<Vec<CosmosMsg>, ServiceError> {
+    // Get the token factory token that represents the liquidity token
+    let token_addr = query_liquidity_token(deps, cfg)?;
+
+    // Query the balance of the account that is going to withdraw
+    let balance_response: BalanceResponse = deps.querier.query_wasm_smart(
+        token_addr.clone(),
+        &cw20::Cw20QueryMsg::Balance {
+            address: cfg.input_addr.to_string(),
+        },
+    )?;
+    if balance_response.balance.is_zero() {
+        return Err(ServiceError::ExecutionError(
+            "Nothing to withdraw".to_string(),
+        ));
+    }
+
+    // Calculate how much we are going to get when we withdraw
+    let withdrawn_assets: Vec<Asset> = deps.querier.query_wasm_smart(
+        cfg.pool_addr.clone(),
+        &valence_astroport_utils::astroport_cw20_lp_token::PoolQueryMsg::Share {
+            amount: balance_response.balance,
+        },
+    )?;
+
+    // Create the withdraw and send messages
+    let withdraw_msg = CosmosMsg::Wasm(cosmwasm_std::WasmMsg::Execute {
+        contract_addr: token_addr.to_string(),
+        msg: to_json_binary(&Cw20ExecuteMsg::Send {
```

Comment:
I don't understand this part .. can you explain how it works ?

Reply:
Yes, in Cw20 we have both Transfer and Send.
- Transfer is similar to a bank send message, just send the CW20 token to an address.
- Send on the other hand, is to send CW20 tokens to a contract and atomically trigger a hook on that contract (in this case, the withdrawliquidity message that astroport defines in its pool contract that I specify below)

In summary, what I'm saying is: Send these tokens (they are cw20, so it's a smart contract execution) to this contract address and immediately trigger the action (hook) that I'm specifying here. Astroport will receive those cw20 tokens and do the action in the contract.

If it's still not clear lmk and we can hop on a quick call.

Reply:
Understood, all clear, thanks !